Add new algorithm: either()

nordlow opened the pull request:
Thanks to D's lazy parameters this function simplify pattern matching:

`either()` similar to Python/Lisp's `or`. Same as D's `any` but returns first parameter value which implicitly convert to true, or last argument otherwise. I guess `any` could be modified but that would cause breakage on its return type.

I've found real use of `either()`in various parsing contexts.

I'm very open to changing the namings if you feel they are misguiding.

Note that `either` currently cannot be `nothrow` because of issue    https://issues.dlang.org/show_bug.cgi?id=12647

DmitryOlshansky commented:
2 things to note: 
- lazy had awful codegen last time I've seen it
- tryEvery is quite non-obvious in how it expects each lazy parameter to affect the same haystack it then restores on failure. A simpler approach is to just make each parameter a callable that accepts `ref S whole`.

DmitryOlshansky commented on std/algorithm/comparison.d:
```diff
@@ -1567,3 +1567,203 @@ unittest
+    {
+        if (result)
+        {
+            results ~= result;
```
These array appends would be highly problematic for the performance of a parser...

nordlow replied:
Done. I return fixed length bool array instead.

burner commented on std/algorithm/comparison.d:
```diff
@@ -1567,3 +1567,203 @@ unittest
+}
+
+/** Returns: true iff all types $(D T) are the same. */
+template allSame(T...)
```
this looks like something that should go into std.traits

nordlow replied:
Moved to end of `std.traits`.

burner commented on std/algorithm/comparison.d:
```diff
@@ -1567,3 +1567,203 @@ unittest
+unittest
+{
+    static assert(allSame!(int, int));
+    static assert(!allSame!(int, double));
```
please test something with > 2 types

nordlow replied:
Done.

burner commented on std/algorithm/comparison.d:
```diff
@@ -1567,3 +1567,203 @@ unittest
+
+    TODO Is inout Conversion!T the correct return value?
+*/
+CommonType!T every(T...)(lazy T a) if (T.length >= 1)
```
if (T.length > 0) or T.length

>= forces me to think, that is not good. > 0 is easier

nordlow replied:
Why is this better?

andralex replied:
Well this is nice for symmetry but doesn't seem awfully useful.

burner commented on std/algorithm/comparison.d:
```diff
@@ -1567,3 +1567,203 @@ unittest
+    assert(every(cast(string)null, "b") == cast(string)null);
+}
+
+version(none) // WARNING disabled because I don't see any use of this for.
```
If you don't see use for it, delete it

nordlow replied:
Done.

burner commented on std/algorithm/comparison.d:
```diff
@@ -1567,3 +1567,203 @@ unittest
+version(none) // WARNING disabled because I don't see any use of this for.
+{
+    /** This overload enables, when possible, lvalue return.
+     */
```
doc doesn't make clear what happens, also unittest don't really help

DmitryOlshansky commented:
@nordlow Please do not merge with master, do:
```sh
 # if upstream is dlang/phobos  remote
git pull --rebase upstream master
```

nordlow commented:
I just realized that it's better for `tryEvery` to return either a

- a `bool` `true upon matching of all arguments, `false` otherwise,
- a `size_t` counter that somehow indicates how many arguments that could be matched or
- a Voldemort struct with `count` member and implicit bool conversion rules that enables the nice idiom

```D
if (const hit = haystack.tryEvery(needles...) 
{
    // do something with `hits.count`
}
```

I prefer the last alternative as it will harmonize with the pending

https://github.com/D-Programming-Language/phobos/pull/3395

What do you guys think?

andralex commented on std/algorithm/comparison.d:
```diff
@@ -1567,3 +1567,156 @@ unittest
+    impossible to mark either() as nothrow. Issue at
+    https://issues.dlang.org/show_bug.cgi?id=12647
+*/
+CommonType!Ts either(Ts...)(lazy Ts a) if (a.length >= 1)
```
Nice, two things:

1. Could you please try whether it's possible to accept/return by reference? I.e. `either(a[0], a[1]) = 1;` should work. But I don't think auto ref cooperates nicely with lazy.

2. A predicate would help, e.g. choose either non-empty string: `auto s = either!(s => !s.empty)(param, "default");`. In fact I think such use would be most useful.

MetaLang replied:
You need to ensure here that `CommonType!Ts` is not `void`.

andralex commented on std/algorithm/comparison.d:
```diff
@@ -1567,3 +1567,156 @@ unittest
+/** This overload enables, when possible, lvalue return.
+    TODO should we limit this template to a.length >= 2
+*/
+auto ref either(Ts...)(ref Ts a) if (a.length >= 1 && allSame!Ts)
```
Ah, hadn't gotten to this point yet. I think this is fine, although there is the freak situation `either(a[0], a[awfullySlowFunction()])` when there's evaluation done for no reason. I think it's safe to gloss over such corner cases.

andralex commented on std/algorithm/comparison.d:
```diff
@@ -1567,3 +1567,156 @@ unittest
+    }
+}
+
+/** This overload enables, when possible, lvalue return.
```
Use /// ditto and one documentation for all overloads.

nordlow replied:
Done.

andralex commented on std/algorithm/comparison.d:
```diff
@@ -1567,3 +1567,156 @@ unittest
                 ));
 }
+
+/** Returns: First Argument (element of $(D a)) whose implicit conversion to
```
Everything code must be enclosed in backquotes or $(D ...). That includes Argument, bool, true, etc.

nordlow replied:
Done.

andralex commented on std/algorithm/comparison.d:
```diff
@@ -1567,3 +1567,156 @@ unittest
+    Similar to behaviour of or operator in dynamic languages such as Lisp's (or
+    a...) and Python's a or ....
+
+    TODO Is inout Conversion!T the correct return value?
```
colon after TODO

nordlow replied:
Done.

andralex commented on std/algorithm/comparison.d:
```diff
@@ -1567,3 +1567,156 @@ unittest
+    values as an array, otherwise restore whole and return null.
+*/
+bool[T.length] tryEvery(S, T...)(ref S whole,
+                                 lazy T parts) if (T.length >= 1)
```
Gave this 30 seconds, couldn't understand what it does (even after looking at examples) or how it can be useful.

nordlow replied:
A better naming might be `skipOverAllOrNone` or perhaps just `skipOverAll` or perhaps just making skipOver `variadic`.

nordlow replied:
Removed it.

andralex commented on std/traits.d:
```diff
@@ -6515,3 +6515,29 @@ unittest
+/**
+Returns: $(D true) iff all types $(D T) are the same.
+*/
+template allSame(T...)
```
OK, apparently we don't have any reasonable way to compute this. Should be named `allSameType` so it's not confused with value-trafficking artifacts.

nordlow replied:
Renamed and put in std.traits.

andralex commented on std/traits.d:
```diff
@@ -6515,3 +6515,29 @@ unittest
+    }
+}
+
+unittest
```
Add /// before unittest to make it a documentation example. All artifacts must have examples.

nordlow replied:
Done

MetaLang commented on std/algorithm/comparison.d:
```diff
@@ -1567,3 +1567,156 @@ unittest
+    }
+    else
+    {
+        return a0 ? a0 : either(a[1 .. $]); // recurse
```
There's no need for this to be recursive, and it will definitely be slower than an iterative approach.

MetaLang commented:
I'd like to propose a different implementation for `either`. I think an iterative approach is preferable for efficiency reasons.

```d
CommonType!Ts either1(Ts...)(lazy Ts a) 
if (a.length >= 1 && !is(CommonType!Ts == void))
{
	foreach (ai; a)
		if (ai) 
			return ai;
	
	return a[$ - 1];
}
```

CyberShadow commented:
So `either`/`every` are `any`/`all` except they return non-bool? Does the distinction make them useful enough to add considering the overlap in function? Can we change `any`/`all` instead?

Yes, `either`/`every` operate on a tuple, and `any`/`all` on a range. But seeing as the functions require that they have a common type, you can use `any`/`all` with `only`, right?

`tryEvery` seems too specialized for Phobos.

DmitryOlshansky commented:
> tryEvery seems too specialized for Phobos.

+1 I'd also point out non-obvious conventions in its usage.

nordlow commented:
Made `either` and `every` iterative instead.

MetaLang commented on std/algorithm/comparison.d:
```diff
@@ -1567,3 +1567,146 @@ unittest
+    foreach (ref e; a[0 .. $ - 1])
+        if (e)
+            return e;
+    return a[$ - 1];
```
This will return a[$ - 1] even if it's false.

nordlow replied:
So which lvalue should be returned if all parameters `bool`-convert to `false`?

MetaLang replied:
It seems that the only sensible thing to do would be to throw an exception or assert, but that seems too heavy for such a small function. Maybe have the user provide a default value? It doesn't really make sense, but I don't see much of an alternative.

nordlow replied:
Python returns the last argument

```Python
>>> False or False
False
>>> 0 or 0
0
>>> 0 or False
False
>>> 1 or False
1
>>> 0 or False
False
>>> False or 0
0
```

MetaLang replied:
Ah, okay.

nordlow replied:
IMO the important thing here is to have the same behaviour for l-values as for r-values. If so I see three alternatives here:

1. Mimic Python's behaviour;
- `either` returns *last* argument upon failure
- `every` returns *first* argument upon failure

2. If default value is supplied return it upon failure for both l-value and r-value overloads.

4. Throw Exceptions

What do you think?

nordlow replied:
Ping. I'm sensing we need a voting on this matter.

andralex replied:
should be fine as is

nordlow commented:
I guess we need some `inout` care for `either` and `every` aswell, right?

MetaLang commented on std/algorithm/comparison.d:
```diff
@@ -1567,3 +1567,156 @@ unittest
+    https://issues.dlang.org/show_bug.cgi?id=12647
+*/
+CommonType!Ts either(Ts...)(lazy Ts a)
+if (a.length >= 1 && !is(CommonType!Ts == void))
```
Better also check that everything can actually convert to bool.

    enum implicitlyConvertsToBool(T) = isImplicitlyConvertible!(T, bool);
    allSatisfy(implicitlyConvertsToBool, Ts);

nordlow replied:
```D
pragma(msg, isImplicitlyConvertible!(bool, typeof(13)));
```

works but

```D
pragma(msg, isImplicitlyConvertible!(bool, typeof("")));
```

doesn't. Why?

MetaLang commented on std/algorithm/comparison.d:
```diff
@@ -1567,3 +1567,156 @@ unittest
+    assert(either(3, 4) == 3);
+    assert(either(0, 4) == 4);
+    assert(either(0, 0) == 0);
+    assert(either("", "a") == "");
```
You should make sure to point out this case as it's not intuitive for someone who doesn't know how D's arrays work.

nordlow commented:
The latest commit contains a todo; there's a problem with the conversion trait. It doesn't work with arrays. Is there an alternative trait?

JakobOvrum commented on std/algorithm/comparison.d:
```diff
@@ -1567,3 +1567,165 @@ unittest
+
+    TODO: Is inout Conversion!T the correct return value?
+
+    NOTE: Lazy parameters are currently, too restrictively, inferred by DMD to
```
It's not just `nothrow`, it's all function attributes.

nordlow replied:
So should I remove this comment for now or tell that now attributes can be inferred?

JakobOvrum commented on std/algorithm/comparison.d:
```diff
@@ -1567,3 +1567,165 @@ unittest
+    allSatisfy!(implicitlyConvertsToBool, Ts))
+{
+    foreach (e; a)
+        if (e)
```
This is actually an explicit conversion to bool, it supports `opCast`. It would be neat if that was supported in the constraint.

nordlow replied:
Could you point any type traits to use for this, please?

nordlow replied:
I guess we should use `ifTestable` here as @andralex noted.

nordlow replied:
@andralex, suggestions on how to restrict parameters to `either` to types that support explicit cast to bool, as @JakobOvrum mentioned?

andralex commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1565,184 @@ unittest
 }
+
+// TODO Why does this return $(D false) when $(D T) is a $(D string) or array: isImplicitlyConvertible!(bool, T);
+enum implicitlyConvertsToBool(T) = isImplicitlyConvertible!(bool, T) || isArray!T;
```
So this is public and undocumented? Also it's a misnomer - arrays are not convertible to bool, they're just if-testable.

nordlow replied:
Se comment above.

andralex commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1565,184 @@ unittest
+
+// TODO Why does this return $(D false) when $(D T) is a $(D string) or array: isImplicitlyConvertible!(bool, T);
+enum implicitlyConvertsToBool(T) = isImplicitlyConvertible!(bool, T) || isArray!T;
+import std.typetuple : allSatisfy;
```
top-level imports should be at the beginning (they have global effect anyway)

nordlow replied:
Done.

andralex commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1565,184 @@ unittest
+enum implicitlyConvertsToBool(T) = isImplicitlyConvertible!(bool, T) || isArray!T;
+import std.typetuple : allSatisfy;
+
+/** Returns: First Argument (element of $(D a)) whose implicit conversion to
```
Why is "argument" capitalized?

andralex commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1565,184 @@ unittest
+    mark $(D either) as $(D nothrow). Issue at
+    https://issues.dlang.org/show_bug.cgi?id=12647
+*/
+CommonType!Ts either(Ts...)(lazy Ts a)
```
This is a nice function to have. It should restrict `a.length > 1` to avoid goofy calls like `x.either` and replace `implicitlyConvertsToBool` with `ifTestable` or simple use a lambda to check whether a type is testable with `if`.

`lazy` is indeed quite inefficient but I see that as an implementation problem.

andralex commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1565,184 @@ unittest
+    foreach (e; a)
+        if (e)
+            return e;
+    return CommonType!Ts.init;
```
Here just return `a[$ - 1]` - it's easy to explain that the list has been exhausted and you return whatever you tried last.

nordlow replied:
@andralex Unless we provided a default value as template parameter right?

andralex commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1565,184 @@ unittest
+}
+
+/// ditto
+auto ref either(Ts...)(ref Ts a)
```
I''m unclear how the two overloads interact. For example, a `ref` may be in fact expensive to compute e.g. `a[fun()]`, but I'd agree that it's a good heuristic to assume that lvalues are generally not expensive to compute.

The question remains what happens when the two overloads conflict?

andralex commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1565,184 @@ unittest
+    static assert(!__traits(compiles, either(1, "a")));
+    static assert(!__traits(compiles, either(1.0, "a")));
+    static assert(!__traits(compiles, either('a', "a")));
+}
```
not seeing any tests for lazy

nordlow replied:
Ahh, good catch. 

I can't figure out how to make both overloads work together.

If remove the ref returning overload lazyness works for 

```D
@safe pure unittest
{
    auto a = 0;
    auto b = 0;
    auto ab = either(++a, ++b); // second argument is not evaluated because first bool-converts to true
    assert(a == 1);
    assert(b == 0);
}
```

But the, of course, 

```D
@safe pure unittest
{
    int x = 1;
    int y = 2;
    either(x, y) = 3;
    assert(x == 3);
    assert(y == 2);
}
```

fails

andralex replied:
First, you need unittests with predicates.

Second, I think the ref-returning either adds too much complication for a few niche use cases. Let's just add the rvalue-returning version.

andralex commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1565,184 @@ unittest
+
+    TODO: Is inout Conversion!T the correct return value?
+*/
+CommonType!Ts every(Ts...)(lazy Ts a)
```
I don't think this is useful, in spite of the completeness argument. There's vast use of `either` as "try this and that and the other until something that works" but the notion of "try this and that and the other and then hey give me the last one" is seldom useful enough to warrant a shortcut.

What /would/ be useful is `either` with a predicate, e.g. "give me the first string longer than four characters: `either!(s => s.length > 4)(s1, s2, "12345")`

nordlow replied:
Removed it.

andralex commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1565,184 @@ unittest
+    static assert(!__traits(compiles, every('a', "a")));
+}
+
+/** Evaluate all $(D parts) possibly digesting $(D whole).
```
What does "to digest" mean in computer science? Only meaning I could think of was the checksum-related meaning.

nordlow replied:
Removed it.

andralex commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1565,184 @@ unittest
+    If all values of $(D parts) implicitly convert to $(D bool) $(D true) return
+    the $(D parts.length), otherwise restore $(D whole) and return -1 number of
+    arguments in $(D parts) that matched.
+*/
```
I protest this algorithm on grounds I made no sense of its description.

nordlow replied:
Removed it.

andralex commented on std/traits.d:
```diff
@@ -6730,3 +6730,29 @@ unittest
+/**
+   Returns: $(D true) iff all types $(D T) are the same.
+*/
+template allSameType(T...)
```
OK

andralex commented on std/traits.d:
```diff
@@ -6730,3 +6730,29 @@ unittest
+        enum bool allSameType = is(T[0] == T[1]) && allSameType!(T[1..$]);
+    }
+}
+
```
Simpler:

    enum allSameType(T) = !T.length || is(T[0] == T[T.length > 1] && allSameType!(T[1 .. $]);

andralex replied:
well I guess not a lot simpler

nordlow replied:
Variant

```D
enum allSameType(T...) = !T.length || (is(T[0] == T[T.length > 1]) && allSameType!(T[1 .. $]));
```

fails to compile 

```D
traits_ex.d(97,71): Error: slice [1..0] is out of range of [0..0]
traits_ex.d(97,71): Error: template instance traits_ex.allSameType!() error instantiating
traits_ex.d(97,71):        instantiated from here: allSameType!int
traits_ex.d(109,19):        instantiated from here: allSameType!(int, int)
traits_ex.d(97,71): Error: template instance traits_ex.allSameType!double error instantiating
traits_ex.d(110,20):        instantiated from here: allSameType!(int, double)
traits_ex.d(111,20): Error: template instance traits_ex.allSameType!(int, int, double) error instantiating
traits_ex.d(183,5): Error: static assert  (is(typeof(__error) == typeof(arr))) is false
```

I stick with my current version for now.

nordlow replied:
BTW: @andralex there's seems to be different views (according to post in the forums) on whether variadic traits should iterate or recurse their CT-parameters. See for example my three different implementions of allSame here: 

https://github.com/nordlow/justd/blob/master/traits_ex.d#L25 (tri-divide-recursion)
https://github.com/nordlow/justd/blob/master/traits_ex.d#L47 (iteration)
https://github.com/nordlow/justd/blob/master/traits_ex.d#L79 (tail-recursion)

Which one should we strive for? Will `static foreach` enable the iterative approach to not require an inner templated function?

andralex replied:
I think you're doing fine here with iteration. In this case you can't halve the interval because by definition you're doing a linear search for the first nonzero.

andralex commented:
I'd say add `either` with this PR, no throwing please, perhaps with optional predicate, and leave `every` and `tryAll` for another fight. Thanks @DmitryOlshansky for pinging me about this. Generally: don't hesitate to email me when I'm the bottleneck of something - the PR traffic is too thick to see everything. Is there a way to filter notifications such that I see those mentioning "@andralex"?

nordlow commented:
Squashed and rebased.

JakobOvrum commented:
@andralex, not sure if you found your answer, but you can view PRs that mention you for specific projects, like so:

https://github.com/D-Programming-Language/phobos/pulls/mentioned/andralex

If you type "is:open is:pr mentions:andralex" in the search field of a project it will automatically redirect you to a link like the above. More search options can be found [here](https://help.github.com/articles/searching-issues/). Personally I also like sorting by "recently updated" instead of the default "newest".

nordlow commented:
One thing remains:

I'm unsure about the naming, definition and usage of

```D
implicitlyConvertsToBool
```

should we call it ```ifTestable``` instead? If so how should it be defined?

nordlow commented:
What about adding a default return argument as a compile-time parameter? Returned when all normal parameters evaluate to `false`?

andralex commented on std/algorithm/comparison.d:
```diff
@@ -17,6 +17,9 @@ $(T2 clamp,
         and $(D cmp("abc", "abc")) is $(D 0).)
+$(T2 either,
+        Return first parameter whose $(D bool)-conversion is $(D true), e.g.
+        $(D either(0, 42, 43)) returns $(D 42).)
```
That wouldn't be the bool conversion. Should be something like: "Return first parameter `pk` that passes an `if (pk)` test, e.g. $(D either(0, 42, 43)) returns $(D 42)."

nordlow replied:
Done.

andralex commented on std/algorithm/comparison.d:
```diff
@@ -1911,3 +1915,97 @@ bool isPermutation(alias pred = "a == b", Range1, Range2)
     ));
 }
+
+enum implicitlyConvertsToBool(T) = isImplicitlyConvertible!(bool, T) || isArray!T;
```
So this should be `ifTestable` and the implementation should just make sure that `if (T()) {}` compiles.

nordlow replied:
Ok. Done. 

I still need to define it as

enum ifTestable(T) = __traits(compiles, { if ((T())) {} }) || isArray!T;

otherwise compilation fails as

```D
comparison.d(1993,18): Error: template std.algorithm.comparison.either cannot deduce function from argument types !()(string, string), candidates are:
comparison.d(1938,15):        std.algorithm.comparison.either(alias pred = (a) => a, Ts...)(lazy Ts a) if (a.length > 1 && !is(CommonType!Ts == void) && allSatisfy!(ifTestable, Ts))
comparison.d(1951,10):        std.algorithm.comparison.either(alias pred = (a) => a, Ts...)(ref Ts a) if (a.length > 1 && allSameType!Ts && allSatisfy!(ifTestable, Ts))
```

Any idea why?

andralex commented on std/algorithm/comparison.d:
```diff
@@ -1911,3 +1915,97 @@ bool isPermutation(alias pred = "a == b", Range1, Range2)
+CommonType!Ts either(alias pred = a => a, Ts...)(lazy Ts a)
+if (a.length > 1 &&
+    !is(CommonType!Ts == void) &&
+    allSatisfy!(implicitlyConvertsToBool, Ts))
```
This condition is wrong because what you need is to make sure that `pred(x)` is if-testable, not `x`.

nordlow replied:
Ok, so I defined it as

```D
enum ifTestable(T, alias pred = a => a) = __traits(compiles, { if ((pred(T()))) {} }) || isArray!T;
```

instead. How do I change the call to

```D
allSatisfy!(ifTestable, Ts)
```

to propagate `pred` into `ifTestable`?

andralex commented on std/algorithm/comparison.d:
```diff
@@ -1911,3 +1915,97 @@ bool isPermutation(alias pred = "a == b", Range1, Range2)
+auto ref either(alias pred = a => a, Ts...)(ref Ts a)
+    if (a.length > 1 &&
+        allSameType!Ts &&
+        allSatisfy!(implicitlyConvertsToBool, Ts))
```
ditto

andralex commented:
Guess this will be ready after one more pass. Famous last words :o).

nordlow commented:
@andralex See the outdate diffs.

JakobOvrum commented on std/algorithm/comparison.d:
```diff
@@ -1911,3 +1915,101 @@ bool isPermutation(alias pred = "a == b", Range1, Range2)
     ));
 }
+
+enum ifTestable(T, alias pred = a => a) = __traits(compiles, { if ((pred(T()))) {} }) || isArray!T;
```
I don't think the `pred` parameter here is ever used.

JakobOvrum replied:
You can partially apply `pred` to pass it to `allSatisfy` with something like this:
```d
enum ifTestable(alias pred, T) = __traits(compiles, { if ((pred(T.init))) {} }) || isArray!T;

template applyLeft(alias higher, appliedArgs...)
{
   enum applyLeft(args...) = higher!(appliedArgs, args);
}

CommonType!Ts either(alias pred = a => a, Ts...)(lazy Ts a)
    if (a.length > 1 &&
        !is(CommonType!Ts == void) &&
        allSatisfy!(applyLeft!(ifTestable, unaryFun!pred), Ts))
{
    ...
}

andralex replied:
I think you want to use `T.init` instead of `T()`. The former works even for types with disabled `this()`. Also, why did you need to special case `isArray!T`? Double also, why are static arrays included? They can't be tested with `if`.

nordlow replied:
Errors as:

```
../../std/meta.d(596,27): Error: template instance F!(const(int)) does not match template declaration ifTestable(alias pred, T)
../../std/meta.d(601,13): Error: template instance std.meta.allSatisfy!(ifTestable, const(int)) error instantiating
comparison.d(1961,9):        instantiated from here: allSatisfy!(ifTestable, const(int), const(int))
../../std/meta.d(596,27): Error: template instance F!(immutable(int)) does not match template declaration ifTestable(alias pred, T)
../../std/meta.d(601,13): Error: template instance std.meta.allSatisfy!(ifTestable, immutable(int)) error instantiating
comparison.d(1961,9):        instantiated from here: allSatisfy!(ifTestable, immutable(int), immutable(int))
../../std/meta.d(596,27): Error: template instance F!int does not match template declaration ifTestable(alias pred, T)
../../std/meta.d(601,13): Error: template instance std.meta.allSatisfy!(ifTestable, int) error instantiating
comparison.d(1961,9):        instantiated from here: allSatisfy!(ifTestable, int, int)
../../std/meta.d(596,27): Error: template instance F!string does not match template declaration ifTestable(alias pred, T)
../../std/meta.d(601,13): Error: template instance std.meta.allSatisfy!(ifTestable, string) error instantiating
comparison.d(1961,9):        instantiated from here: allSatisfy!(ifTestable, string, string)
../../std/meta.d(596,27): Error: template instance F!(immutable(int[])) does not match template declaration ifTestable(alias pred, T)
../../std/meta.d(601,13): Error: template instance std.meta.allSatisfy!(ifTestable, immutable(int[])) error instantiating
comparison.d(1961,9):        instantiated from here: allSatisfy!(ifTestable, immutable(int[]), immutable(int[]))
../../std/meta.d(596,27): Error: template instance F!(int[]) does not match template declaration ifTestable(alias pred, T)
```

andralex replied:
Not sure what's going on here, but instead of `allSatisfy` you may want to use simple recursion: `either` is defined with `n` arguments iff `either` with the `n - 1` tail arguments compiles. We use this trick in a number of places in phobos.

nordlow replied:
Fixed. See below.

JakobOvrum commented on std/algorithm/comparison.d:
```diff
@@ -1911,3 +1915,101 @@ bool isPermutation(alias pred = "a == b", Range1, Range2)
+
+enum ifTestable(T, alias pred = a => a) = __traits(compiles, { if ((pred(T()))) {} }) || isArray!T;
+
+/** Returns: First Parameter $(D p) (element of $(D a)) that passes an $(D if
```
So the whole documentation comment is in the `Returns:` section? That's a little unorthodox...

andralex replied:
Yah, that won't look well. Open with a general description and keep the "Returns:" section small.

andralex replied:
Why is "parameter" title cased?

nordlow replied:
Fixed.

If you want a specific `Returns:` what should I write in it, in order not to repeat things?

andralex commented on std/algorithm/comparison.d:
```diff
@@ -1911,3 +1915,101 @@ bool isPermutation(alias pred = "a == b", Range1, Range2)
+    mark $(D either) as $(D nothrow). Issue at
+    https://issues.dlang.org/show_bug.cgi?id=12647
+*/
+CommonType!Ts either(alias pred = a => a, Ts...)(lazy Ts a)
```
First element cannot be lazy, so please identify in the signature.

```d
CommonType!(T, Ts) either(alias pred = a => a, Ts...)(T first, lazy Ts alternatives)
```

andralex commented on std/algorithm/comparison.d:
```diff
@@ -1911,3 +1915,101 @@ bool isPermutation(alias pred = "a == b", Range1, Range2)
+    foreach (ref e; a[0 .. $ - 1])
+        if (predicate(e))
+            return e;
+    return a[$ - 1];
```
I think this evaluates the last argument again, so a better form is:

```D
foreach (e; a[0 .. $ - 1])
{
    if (predicate(e)) return e;
}
return a[$ - 1]();
```

nordlow replied:
Done.

andralex commented:
Couple more nits and we're ready to go... again :o).

nordlow commented:
After my latest changes in the first `either`-overload we need to restrict it to not allow it when all variables are l-values in order for the second overload to kick in. How do I do that?

andralex commented:
Regarding lvalues: I'd say let's just forgo them.

nordlow commented:
Ok, removed it.

nordlow commented:
Updated to use `pred(T.init)` instead of `pred(T())` and removed array-check from `ifTestable()`.

I'll look into the problem with pred evaluation aswell.

nordlow commented:
- Moved `ifTestable` to `std.traits`
- Added static unittests for `ifTestable` to prove that pred is indeed evaluated

andralex commented on std/algorithm/comparison.d:
```diff
@@ -1911,3 +1915,87 @@ bool isPermutation(alias pred = "a == b", Range1, Range2)
+    if (alternatives.length >= 1 &&
+        !is(CommonType!(T, Ts) == void) &&
+        ifTestable!T &&
+        allSatisfy!(ifTestable, Ts))
```
Just replace these two lines with `allSatisfy!(ifTestable, T, Ts)`

andralex commented:
One last nit. Then I preapprove (it sounds... oxymoronic).

nordlow commented:
Done!

andralex commented:
Auto-merge toggled on

andralex commented:
thx!

nordlow commented:
Great!

nordlow commented:
Just realized that we can make use of `typeof(return)` in

```D
CommonType!(T, Ts) either(alias pred = a => a, T, Ts...)(T first, lazy Ts alternatives)
    if (alternatives.length >= 1 &&
        !is(typeof(return) == void) &&
        allSatisfy!(ifTestable, T, Ts))
```

Would that be preferrable in other PRs?

JakobOvrum commented:
This *still* isn't passing the predicate to `ifTestable`, despite my comments...

nordlow commented:
Should we revert and fix, @andralex ?

andralex commented:
@nordlow just create a new PR. Thanks @JakobOvrum

nordlow commented:
Ok.
